File: example/neural-style/end_to_end/boost_train.py
import sys
sys.path.insert(0, "../../mxnet/python")

import mxnet as mx
import numpy as np

import basic
import data_processing
import gen_v3
import gen_v4

# params
vgg_params = mx.nd.load("./vgg19.params")
style_weight = 1.2
content_weight = 10
dshape = (1, 3, 384, 384)
clip_norm = 0.05 * np.prod(dshape)
model_prefix = "v3"
ctx = mx.gpu(0)

# init style
style_np = data_processing.PreprocessStyleImage("../starry_night.jpg", shape=dshape)
style_mod = basic.get_style_module("style", dshape, ctx, vgg_params)
style_mod.forward(mx.io.DataBatch([mx.nd.array(style_np)], [0]), is_train=False)
style_array = [arr.copyto(mx.cpu()) for arr in style_mod.get_outputs()]
del style_mod

# content
content_mod = basic.get_content_module("content", dshape, ctx, vgg_params)

# loss
loss, gscale = basic.get_loss_module("loss", dshape, ctx, vgg_params)
extra_args = {"target_gram_%d" % i : style_array[i] for i in range(len(style_array))}
loss.set_params(extra_args, {}, True, True)
grad_array = []
for i in range(len(style_array)):
    grad_array.append(mx.nd.ones((1,), ctx) * (float(style_weight) / gscale[i]))
grad_array.append(mx.nd.ones((1,), ctx) * (float(content_weight)))

# generator
gens = [gen_v4.get_module("g0", dshape, ctx),
        gen_v3.get_module("g1", dshape, ctx),
        gen_v3.get_module("g2", dshape, ctx),
        gen_v4.get_module("g3", dshape, ctx)]
for gen in gens:
    gen.init_optimizer(
        optimizer='sgd',
        optimizer_params={
            'learning_rate': 1e-4,
            'momentum' : 0.9,
            'wd': 5e-3,
            'clip_gradient' : 5.0
        })


# tv-loss
def get_tv_grad_executor(img, ctx, tv_weight):
    """create TV gradient executor with input binded on img
    """
    if tv_weight <= 0.0:
        return None
    nchannel = img.shape[1]
    simg = mx.sym.Variable("img")
    skernel = mx.sym.Variable("kernel")
    channels = mx.sym.SliceChannel(simg, num_outputs=nchannel)
    out = mx.sym.Concat(*[
        mx.sym.Convolution(data=channels[i], weight=skernel,
                           num_filter=1,
                           kernel=(3, 3), pad=(1,1),
                           no_bias=True, stride=(1,1))
        for i in range(nchannel)])
    kernel = mx.nd.array(np.array([[0, -1, 0],
                                   [-1, 4, -1],
                                   [0, -1, 0]])
                         .reshape((1, 1, 3, 3)),
                         ctx) / 8.0
    out = out * tv_weight
    return out.bind(ctx, args={"img": img,
                               "kernel": kernel})
tv_weight = 1e-2

start_epoch = 0
end_epoch = 3


# data
import os
import random
import logging

data_root = "../data/"
file_list = os.listdir(data_root)
num_image = len(file_list)
logging.info("Dataset size: %d" % num_image)


# train

for i in range(start_epoch, end_epoch):
    random.shuffle(file_list)
    for idx in range(num_image):
        loss_grad_array = []
        data_array = []
        path = data_root + file_list[idx]
        content_np = data_processing.PreprocessContentImage(path, min(dshape[2:]), dshape)
        data = mx.nd.array(content_np)
        data_array.append(data)
        # get content
        content_mod.forward(mx.io.DataBatch([data], [0]), is_train=False)
        content_array = content_mod.get_outputs()[0].copyto(mx.cpu())
        # set target content
        loss.set_params({"target_content" : content_array}, {}, True, True)
        # gen_forward
        for k in range(len(gens)):
            gens[k].forward(mx.io.DataBatch([data_array[-1]], [0]), is_train=True)
            data_array.append(gens[k].get_outputs()[0].copyto(mx.cpu()))
            # loss forward
            loss.forward(mx.io.DataBatch([data_array[-1]], [0]), is_train=True)
            loss.backward(grad_array)
            grad = loss.get_input_grads()[0]
            loss_grad_array.append(grad.copyto(mx.cpu()))
        grad = mx.nd.zeros(data.shape)
        for k in range(len(gens) - 1, -1, -1):
            tv_grad_executor = get_tv_grad_executor(gens[k].get_outputs()[0],
                    ctx, tv_weight)
            tv_grad_executor.forward()

            grad[:] += loss_grad_array[k] + tv_grad_executor.outputs[0].copyto(mx.cpu())
            gnorm = mx.nd.norm(grad).asscalar()
            if gnorm > clip_norm:
                grad[:] *= clip_norm / gnorm

            gens[k].backward([grad])
            gens[k].update()
        if idx % 20 == 0:
            logging.info("Epoch %d: Image %d" % (i, idx))
            for k in range(len(gens)):
                logging.info("Data Norm :%.5f" %\
                        (mx.nd.norm(gens[k].get_input_grads()[0]).asscalar() / np.prod(dshape)))
        if idx % 1000 == 0:
            for k in range(len(gens)):
                gens[k].save_params("./model/%d/%s_%04d-%07d.params" % (k, model_prefix, i, idx))

File: scala-package/examples/src/main/scala/ml/dmlc/mxnetexamples/neuralstyle/end2end/BoostTrain.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ml.dmlc.mxnetexamples.neuralstyle.end2end

import org.slf4j.LoggerFactory
import org.kohsuke.args4j.{CmdLineParser, Option}
import scala.collection.JavaConverters._
import ml.dmlc.mxnet.NDArray
import ml.dmlc.mxnet.Shape
import ml.dmlc.mxnet.Context
import ml.dmlc.mxnet.DataBatch
import ml.dmlc.mxnet.Symbol
import ml.dmlc.mxnet.Executor
import ml.dmlc.mxnet.optimizer.SGD
import java.io.File
import javax.imageio.ImageIO
import scala.util.Random
import ml.dmlc.mxnet.optimizer.Adam

/**
 * @author Depeng Liang
 */
object BoostTrain {

  private val logger = LoggerFactory.getLogger(classOf[BoostTrain])

  def getTvGradExecutor(img: NDArray, ctx: Context, tvWeight: Float): Executor = {
    // create TV gradient executor with input binded on img
    if (tvWeight <= 0.0f) null

    val nChannel = img.shape(1)
    val sImg = Symbol.Variable("img")
    val sKernel = Symbol.Variable("kernel")
    val channels = Symbol.SliceChannel()(sImg)(Map("num_outputs" -> nChannel))
    val out = Symbol.Concat()((0 until nChannel).map { i =>
      Symbol.Convolution()()(Map("data" -> channels.get(i), "weight" -> sKernel,
                    "num_filter" -> 1, "kernel" -> "(3,3)", "pad" -> "(1,1)",
                    "no_bias" -> true, "stride" -> "(1,1)"))
    }.toArray: _*)() * tvWeight
    val kernel = {
      val tmp = NDArray.empty(Shape(1, 1, 3, 3), ctx)
      tmp.set(Array[Float](0, -1, 0, -1, 4, -1, 0, -1, 0))
      tmp / 8.0f
    }
    out.bind(ctx, Map("img" -> img, "kernel" -> kernel))
  }

  def main(args: Array[String]): Unit = {
    val stin = new BoostTrain
    val parser: CmdLineParser = new CmdLineParser(stin)
    try {
      parser.parseArgument(args.toList.asJava)
      assert(stin.dataPath != null
          && stin.vggModelPath != null
          && stin.saveModelPath != null
          && stin.styleImage != null)
      // params
      val vggParams = NDArray.load2Map(stin.vggModelPath)
      val styleWeight = 1.2f
      val contentWeight = 10f
      val dShape = Shape(1, 3, 384, 384)
      val clipNorm = 0.05f * dShape.product
      val modelPrefix = "v3"
      val ctx = if (stin.gpu == -1) Context.cpu() else Context.gpu(stin.gpu)

      // init style
      val styleNp = DataProcessing.preprocessStyleImage(stin.styleImage, dShape, ctx)
      var styleMod = Basic.getStyleModule("style", dShape, ctx, vggParams)
      styleMod.forward(Array(styleNp))
      val styleArray = styleMod.getOutputs().map(_.copyTo(Context.cpu()))
      styleMod.dispose()
      styleMod = null

      // content
      val contentMod = Basic.getContentModule("content", dShape, ctx, vggParams)

      // loss
      val (loss, gScale) = Basic.getLossModule("loss", dShape, ctx, vggParams)
      val extraArgs = (0 until styleArray.length)
                                  .map( i => s"target_gram_$i" -> styleArray(i)).toMap
      loss.setParams(extraArgs)
      var gradArray = Array[NDArray]()
      for (i <- 0 until styleArray.length) {
        gradArray = gradArray :+ (NDArray.ones(Shape(1), ctx) * (styleWeight / gScale(i)))
      }
      gradArray = gradArray :+ (NDArray.ones(Shape(1), ctx) * contentWeight)

      // generator
      val gens = Array(
          GenV4.getModule("g0", dShape, ctx),
          GenV3.getModule("g1", dShape, ctx),
          GenV3.getModule("g2", dShape, ctx),
          GenV4.getModule("g3", dShape, ctx)
      )
      gens.foreach { gen =>
        val opt = new SGD(learningRate = 1e-4f,
                          momentum = 0.9f,
                          wd = 5e-3f,
                          clipGradient = 5f)
        gen.initOptimizer(opt)
      }

      var filelist = new File(stin.dataPath).list().toList
      val numImage = filelist.length
      logger.info(s"Dataset size: $numImage")

      val tvWeight = 1e-2f

      val startEpoch = 0
      val endEpoch = 3

      for (k <- 0 until gens.length) {
        val path = new File(s"${stin.saveModelPath}/$k")
        if (!path.exists()) path.mkdir()
      }

      // train
      for (i <- startEpoch until endEpoch) {
        filelist = Random.shuffle(filelist)
        for (idx <- filelist.indices) {
          var dataArray = Array[NDArray]()
          var lossGradArray = Array[NDArray]()
          val data =
            DataProcessing.preprocessContentImage(s"${stin.dataPath}/${filelist(idx)}", dShape, ctx)
          dataArray = dataArray :+ data
          // get content
          contentMod.forward(Array(data))
          // set target content
          loss.setParams(Map("target_content" -> contentMod.getOutputs()(0)))
          // gen_forward
          for (k <- 0 until gens.length) {
            gens(k).forward(dataArray.takeRight(1))
            dataArray = dataArray :+ gens(k).getOutputs()(0)
            // loss forward
            loss.forward(dataArray.takeRight(1))
            loss.backward(gradArray)
            lossGradArray = lossGradArray :+ loss.getInputGrads()(0)
          }
          val grad = NDArray.zeros(data.shape, ctx)
          for (k <- gens.length - 1 to 0 by -1) {
            val tvGradExecutor = getTvGradExecutor(gens(k).getOutputs()(0), ctx, tvWeight)
            tvGradExecutor.forward()
            grad += lossGradArray(k) + tvGradExecutor.outputs(0)
            val gNorm = NDArray.norm(grad)
            if (gNorm.toScalar > clipNorm) {
              grad *= clipNorm / gNorm.toScalar
            }
            gens(k).backward(Array(grad))
            gens(k).update()
            gNorm.dispose()
            tvGradExecutor.dispose()
          }
          grad.dispose()
          if (idx % 20 == 0) {
            logger.info(s"Epoch $i: Image $idx")
            for (k <- 0 until gens.length) {
              val n = NDArray.norm(gens(k).getInputGrads()(0))
              logger.info(s"Data Norm : ${n.toScalar / dShape.product}")
              n.dispose()
            }
          }
          if (idx % 1000 == 0) {
            for (k <- 0 until gens.length) {
              gens(k).saveParams(
                  s"${stin.saveModelPath}/$k/${modelPrefix}_" +
                  s"${"%04d".format(i)}-${"%07d".format(idx)}.params")
            }
          }
          data.dispose()
        }
      }
    } catch {
      case ex: Exception => {
        logger.error(ex.getMessage, ex)
        parser.printUsage(System.err)
        sys.exit(1)
      }
    }
  }
}

class BoostTrain {
  @Option(name = "--data-path", usage = "the input train data path")
  private val dataPath: String = null
  @Option(name = "--vgg--model-path", usage = "the pretrained model to use: ['vgg']")
  private val vggModelPath: String = null
  @Option(name = "--save--model-path", usage = "the save model path")
  private val saveModelPath: String = null
  @Option(name = "--style-image", usage = "the style image")
  private val styleImage: String = null
  @Option(name = "--gpu", usage = "which gpu card to use, default is -1, means using cpu")
  private val gpu: Int = -1
}
